[Feature] Expert Parallelism Load Balancer (EPLB)

abmfy opened the pull request:
This PR introduces support for **dynamic load balancing** in **expert parallelism (EP)** for the deployment of **Mixture-of-Experts (MoE)** models.

Dynamic load balancing is essential for auxiliary-loss-free MoE models, such as the **DeepSeek-V3/R1** series. This feature enables dynamic rearrangement of experts across different ranks/nodes to achieve better load balance during inference.

Additionally, this PR introduces support for **redundant experts**, allowing each routed expert to maintain multiple parameter copies distributed across different ranks. This further improves expert load balancing.

# Running

To try out EPLB, enable it with the following options:

```bash
--enable-eplb
--num-redundant-experts 32
--eplb-window-size 1000
--eplb-step-interval 3000
```

You should see a log message indicating that EPLB is enabled, as well as periodic logs showing the rearrangement of experts.

# Compatibility

Currently, we support DeepSeek-V2, V3, and R1 models with FP8 quantization. However, this PR has been designed with generality in mind, so extending support to other MoE models or quantization methods should be straightforward.

**Adding model support:**

To add support for a new model, implement the `MixtureOfExperts` protocol. In essence, you’ll need to:

- Expose relevant MoE configurations.
- Provide access to the expert weights that need to be shuffled.
- Forward EPLB-related information into the `FusedMoE` layer.

**Note:** Pay close attention to the weight-loading logic. With redundant experts, you’ll need to handle additional complexity to ensure weights are loaded correctly. The `expert_params_mapping` returned by `FusedMoE` reflects the presence of redundant experts, but you may need to implement some nontrivial adjustments in the model class to prevent breaking the weight-loading process.

You can refer to the implementation changes in [deepseek_v2.py](https://github.com/vllm-project/vllm/pull/18343/files#diff-420f1cd67991a63cb419ca0e00e6f42cbe825864d0541e0662eeed2f9ddbd021).

**Adding quantization support:**

Adding quantization support should be straightforward, as it mainly involves forwarding the necessary arguments.

See the changes in [fp8.py](https://github.com/vllm-project/vllm/pull/18343/files#diff-5511bfcc9c53f7d96517ad43e4087f6777bef21302da983f42cafae40a866644) for reference.

**We welcome contributions to help add support for additional models and quantization methods!**

# To-Dos

**To-Do List for this PR:**
- [x] Implement replicated experts in fused MoE operations
- [x] Monitor expert balancedness in metrics
- [x] Remove magic numbers
- [x] Allow turning off monitoring since it brings some overhead

**Long-term To-Do List (should be done in other PRs):**
- Model Execution
  - [ ] When using `FusedMoEModularKernel`, we can directly use the load metrics returned by `FusedMoEPrepareAndFinalize`, instead of calculating them inside expert selection. We're not doing this since not all code paths are using `FusedMoEModularKernel` now
- EPLB Algorithm
  - [ ] Add other rebalancing strategies, e.g. rebalance when balancedness falls below some threshold
  - [ ] Consider treating differently for prefill and decode nodes in the rearrangement algorithm
- EPLB Execution
  - [ ] Parallelize the rearrangement algorithm (calculating new expert mapping, not the communication)
  - [ ] Shuffle one layer at once and use multiple steps, to lower the impact on inter-token latency
  - [ ] Investigate should we pre-allocate expert weight buffer used for transferring
  - [ ] Take locality into consideration in expert weight transmission, e.g. prioritize transferring to GPUs on the same node
- Compatibility
  - [ ] Add support for DeepSeek Multi-Token Prediction (MTP) layers
  - [ ] Add support for two-batch overlap (#18415)
  - [ ] Add support for other MoE models, e.g. Llama 4, Qwen3
  - [ ] Add support for other quantization methods
- API
  - [ ] Group EPLB configs together

abmfy committed:
[Feature] Core EPLB algorithm

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Feature] Register expert weights for DeepSeek MoE

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Chore] Rename EPLB rebalance algo module name

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Feature] Store EPLB states in model runner

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Feature] EPLB rearrangement execution

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Feature] Add expert load metrics collection during forward

WIP, design choices not finalized.

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Feature] Rearrange experts after a preset step interval

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
Merge branch 'main' into eplb

Signed-off-by: Bowen Wang <abmfy@icloud.com>

youkaichao commented on vllm/distributed/eplb/states.py:
```diff
@@ -0,0 +1,257 @@
+    """
+    Mapping from physical experts to logical experts.
+
+    Shape: (num_moe_layers, num_physical_experts)
```
can we run the algorithm per-layer? what's the performance impact?

youkaichao commented on vllm/model_executor/models/deepseek.py:
```diff
@@ -160,13 +203,28 @@ def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
+                    local_mask] - self.physical_expert_start
+                counts = torch.bincount(
+                    local_indices, minlength=self.n_local_physical_experts)
+                expert_load_pass[self.layer_idx] += counts
```
since this is per-layer statistic, maybe we can just store it in `FusedMoE` object?

abmfy replied:
I think that's a better idea, and generalizes to other MoE models too.

abmfy replied:
Regarding the previous comment, I think it's compatible to store per-layer load metrics here by passing in a sliced view of the load matrix in `EplbState`

abmfy committed:
[Feature] Use unified `FusedMoE` in DeepSeek-V3/R1

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Bugfix] Copy expert mappings after rearrangement

Signed-off-by: Bowen Wang <abmfy@icloud.com>

wpc commented on vllm/model_executor/models/deepseek.py:
```diff
@@ -160,13 +203,28 @@ def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
+            if local_mask.any():
+                local_indices = topk_ids[
+                    local_mask] - self.physical_expert_start
+                counts = torch.bincount(
```
I think this will introduce gpu to cpu sync and hurt performance? Also do we use cuda graph in vllm? If so local_mask.any and bincount probably does not work with cudagraph recording?

wpc replied:
Probably better to passing down and record in fused_moe kernel

abmfy replied:
Yes, I’ve been transitioning to a new design where the expert load metrics are recorded in`FusedMoEModularKernel`, since the expert counts are already calculated in  https://github.com/vllm-project/vllm/blob/1f079540db5f1080a2f61a730da50d3009934c5a/vllm/model_executor/layers/fused_moe/modular_kernel.py#L338-L340

y-null commented on vllm/distributed/eplb/rebalance_execute.py:
```diff
@@ -0,0 +1,280 @@
+                weight[dst].copy_(buffer[dst])
+        else:
+            expert = new_indices[local2global(dst)]
+            src = experts_recv_loc[expert]
```
hi, is there any difference between src and dst here? It seems that they are the same from the initialization of experts_rec_loc

abmfy committed:
[Chore] Move implementations to `deepseek_v2.py`

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Chore] Remove expert load stats from forward context

Moved into `FusedMoE` layers

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Feature] Weight loading for redundant experts

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Feature] Expert replica selection and load metrics recording

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Feature] Map logical experts in weight loading

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Bugfix] Use `scatter_add_` instead of `bincount` for compile

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Bugfix] Add EPLB args in `EngineArgs`

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Bugfix] Sum up steps on EPLb rearrange

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Bugfix] Collect expert weights into a list

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Bugfix] Fix typo in assertion

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Bugfix] Pad `log2phy` magging in rebalance algo

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Bugfix] Fix EP group in `DeepseekV2MoE`

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Refactor] Use local physical ids in expert load collection

This is to facilitize the calculation of the load of a rank

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Bugfix] Map physical id before recording expert load metrics

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Perf] Reduce overhead of expert load recording

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
Merge branch 'main' into eplb

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy commented:
> A couple things:
> 
> 1. Can we extract Hierarchical Load Balancing and Global Load Balancing to a config flag?
> 2. The deepseek v3 technical paper (https://arxiv.org/pdf/2412.19437) introduced the dynamic
>    redundancy concept, where more experts can be loaded into GPU HBM, but not activated. Do we plan to add this feature? If we do, we'll need more config parameters to set # of experts to load and # of experts to activate.

Thanks for commenting!

1. Yes, DeepSeek suggested using hierarchical load balancing for pre-fill nodes and global for decode nodes, so I think we could add a flag to switch between two strategies. 
2. Sure, redundant experts have already been added and a `num_extra_experts` option in engine arguments has been added to turn on that feature.

    Regarding the number of experts to activate, that’s part of the model structure and is in the model config.

lhsjohn commented:
Will the load statistics here bring additional performance consumption? Is there a better optimization method?

abmfy commented:
> Will the load statistics here bring additional performance consumption? Is there a better optimization method?

I tested the PyTorch implementation, and the overhead is only about 1/5 of the time taken by the`grouped_topk` operation. Given the improvements brought by load balancing, this overhead seems acceptable. Further optimizations could involve fusing the load statistics collection and the`grouped_topk` operation into a single kernel.

abmfy committed:
[Bugfix] Step EPLB state in dummy run to avoid blocking DP

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Feature] Do not record expert loads for dummy batches

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Bugfix] Collect expert weights after weight post-processing

Instead collecting expert weights during weight loading,
we collect it later after the post-processing, since some
processing like quantization will make the original weights
unavailable.

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Bugfix] Fix weight loading of replica experts

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy commented:
> Hello! That's a great work, indeed! Thank you for this! There are issues in your implementation, however:
> 
> 1. While moving the expert load statistics gathering algorithm from deepseek_v2.py to layer.py, you've introduced temporary restrictions on the parameter values, namely:
>    
>    1. (num_experts + n) % ep_size = 0, where num_experts = 256 for the DeepSeek model, and n is a number of redundant experts (https://github.com/abmfy/vllm/blob/eplb/vllm/model_executor/layers/fused_moe/layer.py#L771).
>    2. Besides, you're using a list comprehension [weight.view(self.local_num_experts, -1) for weight in weights] at https://github.com/abmfy/vllm/blob/eplb/vllm/model_executor/layers/fused_moe/layer.py#L1198, which makes another restriction num_experts % y = 0, where y = (num_experts + n) // ep_size. This set of restrictions makes varying n almost impossible by now (the only reasonable pair of parameters, which satisfy both the restrictions, excluding n = 0, is ep_size = 5, n = 64).
> 
> Could you please prioritize getting rid of these restrictions if it's posible?
> 
> 2. There are two small bugs in the code:
>    
>    1. in the vllm/vllm/distributed/eplb/states.py file, Ll. 126-129 - the tensors, which are using as indexes, need to be converted to integer by logical_idx.int() and logical_replica_count[logical_idx.int()].int(), for instance;
>    2. in the /vllm/vllm/v1/worker/gpu_model_runner.py file a class variable  eplb_state: EplbState (L. 82) needs to be initialized by None (eplb_state: EplbState = None), for example, and to be addressed by the classname instead of self throughout the code (Ll. 1460, 1586):
>       if self.parallel_config.enable_eplb:
>       GPUModelRunner.eplb_state.step(self.model) -- L. 1460,
>       GPUModelRunner.eplb_state = EplbState.build(	-- L. 1586
>       self.model,
>       self.device,
>       self.parallel_config,
>       )
> 
> Hope my notices were useful.

Thanks for your comments!

1. ii. No restrictions are introduced here, as the first dimension of `weight` corresponds to `local_num_experts`. In the EP setting, we only create weights for `local_num_experts`, rather than for all routed experts.

2.  
   i. `physical_to_logical_map` is an integer tensor, so its view, `logical_idx`, is also of integer type—likewise for `logical_replica_count`.  
   ii. No, it's not a class variable; I’m simply using it as a type hint. I’ll look into whether there are better ways to provide this type hint later.

Let me know if anything needs clarification or further discussion.

abmfy committed:
Merge branch 'main' into eplb

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
Merge branch 'main' into eplb

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Bugfix] Remove `e_score_correction_bias` in expert weights

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Bugfix] Fix shapes and dtypes in `FusedMoE`

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
Merge branch 'main' into eplb

Signed-off-by: Bowen Wang <abmfy@icloud.com>

ztxdcyy commented:
Hi @abmfy 👋,

First, thank you for your clever algorithm design and continuous contributions! 🙌 I have two questions about the implementation:

1️⃣ EPLB Rearrangement Trigger
In vllm/distributed/eplb/states.py, do we currently have (or plan to implement) a mechanism to trigger rearrange operations based on balancedness metrics? I'd love to help discuss the algorithm design or contribute code for this feature!

2️⃣ Micro-batch Overlap
In your TODOs, you mentioned "Add support for micro-batch overlap" - could you elaborate on which components this would overlap? (I'm guessing it's between EPLB scheduling and MoE inference, but wanted to confirm)

I'm excited to collaborate on these! Looking forward to your thoughts 💡

abmfy committed:
[Feature] Disable EPLb step during profile run

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy commented:
> Hi @abmfy 👋,
> 
> First, thank you for your clever algorithm design and continuous contributions! 🙌 I have two questions about the implementation:
> 
> 1️⃣ EPLB Rearrangement Trigger In vllm/distributed/eplb/states.py, do we currently have (or plan to implement) a mechanism to trigger rearrange operations based on balancedness metrics? I'd love to help discuss the algorithm design or contribute code for this feature!
> 
> 2️⃣ Micro-batch Overlap In your TODOs, you mentioned "Add support for micro-batch overlap" - could you elaborate on which components this would overlap? (I'm guessing it's between EPLB scheduling and MoE inference, but wanted to confirm)
> 
> I'm excited to collaborate on these! Looking forward to your thoughts 💡

Hi @ztxdcyy, thanks for your comments!

1. Yes, I've been trying to collect balancedness metrics at every model forward, but the overhead is relatively high, so the benefit isn't very clear—especially compared to the simpler strategy of skipping metric collection and just performing rearrangement every fixed number of steps. I'm considering switching to collecting metrics every *N* steps instead of every single one. Definitely open to other suggestions if you have any!
2. Sorry if the wording was a bit unclear. What I meant was adding compatibility *between* EPLB and DBO. The related DBO PR is #18415 if you're interested. I'll update the phrasing to make it less confusing.

Thanks again for your valuable feedback!

abmfy committed:
[Bugfix] Synchronize CUDA before shuffling layer to avoid hang

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
Merge branch 'main' into eplb

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
Merge branch 'eplb-graph' into eplb

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Style] Rename module `eplb.states` to `eplb.eplb_state`

This is because the logger will display the filename.
[INFO] [states.py] is kind of confusing.

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Feature] Run a dummy rearrangement during profile run for CUDA graphs

I'm actually not very sure why we need this.
When CUDA graph in on, we need to perform a communication during the profile run,
maybe to init the NCCL buffer (even though this is outside the graph itself) to
comply with memory constraints of CUDA graphs.

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
Merge branch 'eplb-graph' into eplb

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Feature] Constrain EPLB to main models

Otherwise, if EPLB is enabled, `FusedMoE` in MTP model
will also try to use expert mapping and record load metrics,
which is not wanted.

Later work should consider adding support for `EplbState`
to register multiple models, hence we can do EPLB on both
the main model and the MTP model.

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Refactor] Move out `EplbState` in model runner from classvars

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy commented:
> In my test, it takes seconds to run eplb algorithm. When each GPUModelRunner runns into EplbState.rearrange, why not let each to compute several layers by calling rebalance_experts and all gather the finally results to reduce algorithm latency.
> 
> Further, I found eplb algorithm can generate phy2log with duplicate hot expert ids on same rank. Whether the shuffle_layer for expert weight movement can handle such situation?
> 
> [deepseek-ai/EPLB#22](https://github.com/deepseek-ai/EPLB/issues/22)
> 
> @abmfy

Hi @SoonyangZhang, thank you for your insightful comments!

The main bottleneck lies in the subsequent transmission of model weights. Since the expert mappings themselves are relatively small, I’m concerned that the reduced computation may not offset the increased communication overhead—especially when the number of ranks is large.

I do have two plans in mind to reduce the latency:
	1.	Optimize the algorithm itself to reduce computation overhead;
	2.	Perform EPLB on only one layer per forward pass, to amortize the latency across multiple steps.

I’ll also explore whether distributing the expert mapping computation could improve performance.

Thanks again for your valuable suggestions!

abmfy commented:
> In my test, it takes seconds to run eplb algorithm. When each GPUModelRunner runns into EplbState.rearrange, why not let each to compute several layers by calling rebalance_experts and all gather the finally results to reduce algorithm latency.
> 
> Further, I found eplb algorithm can generate phy2log with duplicate hot expert ids on same rank. Whether the shuffle_layer for expert weight movement can handle such situation?
> 
> [deepseek-ai/EPLB#22](https://github.com/deepseek-ai/EPLB/issues/22)
> 
> @abmfy

Hi @SoonyangZhang,

After some initial implementation, I found calculating a portion of layers on each rank and then gather together indeed reduces the overhead of the algorithm a lot. We should move on to that, but let's open up another PR for that.

Would you have time to do this optimization? If not, I'll make some time to implement it later.

Again, thanks for your valuable idea!

SoonyangZhang commented:
As for running several layers of eplb on each rank, your implementation is already very well. 

In my test, eplb algorithm indeed generate duplicate expert ids on same rank. Once this happens and the weight movement can not cover this, the accuracy of the model will be impacted.
[200, 226, 236, 236, 12, 204, 156, 229, 184], 
after the expert weight is transferred, we could copy the weight of the first 236 to the second 236. 
@abmfy

WoosukKwon commented on vllm/config.py:
```diff
@@ -1762,6 +1762,15 @@ class ParallelConfig:
+    """Enable expert parallelism load balancing for MoE layers."""
+    num_extra_experts: int = 0
+    """Number of redundant experts to use for expert parallelism."""
+    eplb_window_size: int = 1000
+    """Window size for expert load recording."""
+    eplb_step_interval: int = 3000
+    """Interval for rearranging experts in expert parallelism."""
```
Again, please comment that the last `eplb_window_size` steps will be used for rebalancing.

WoosukKwon commented on vllm/distributed/eplb/eplb_state.py:
```diff
@@ -0,0 +1,346 @@
+        # TODO(bowen): Treat differently for prefill and decode nodes
+        num_replicas = model.num_physical_experts
+        num_groups = model.num_expert_groups
+        # TODO(bowen): Remove magic numbers
+        num_nodes = (ep_group.size() + 7) // 8
+        num_gpus = ep_group.size()
```
This needs to be fixed as we might have 4 gpus (instead of 8) per node.

WoosukKwon commented on vllm/distributed/eplb/eplb_state.py:
```diff
@@ -0,0 +1,346 @@
+        # `num_tokens`: (num_moe_layers,)
+        num_tokens = self.expert_load_pass.sum(dim=-1)
+
+        # Collect load metrics from all ranks
+        ep_group = get_ep_group().device_group
+        num_tokens_list = [
+            torch.empty_like(num_tokens) for _ in range(ep_group.size())
+        ]
+        all_gather(num_tokens_list, num_tokens, group=ep_group)
+        # Stack to get (num_ranks, num_moe_layers)
+        num_tokens_per_rank = torch.stack(num_tokens_list).float()
+
+        # Compute balancedness ratio:
+        # for each layer:
+        #   (mean load across ranks) / (max load across ranks)
+        avg_tokens_tensor = num_tokens_per_rank.mean(dim=0).sum(dim=0)
+        max_tokens_tensor = num_tokens_per_rank.max(dim=0).values.sum(dim=0)
+
+        # Just to make type checker happy
+        tokens_tensors: list[float] = torch.stack(
+            [avg_tokens_tensor, max_tokens_tensor]).tolist()
+        avg_tokens, max_tokens = tokens_tensors
+        balancedness = avg_tokens / max_tokens if max_tokens > 0 else 0.0
```
As we discussed offline, please skip this logic for normal cases.

WoosukKwon reviewed (commented):
@abmfy Great work on the PR! I find it clean and well-organized. I also appreciate the thoughtful handling of tricky edge cases. A few follow-ups:

1. As we discussed offline last Friday, please add clarifying comments and rename a few variables for readability.
2. The rebalancing algorithm from the DeepSeek OSS code can be further optimized; let’s do that in the next PR.
3. Some “mystery” hacks remain (e.g., the is_profile flag). Let’s investigate and remove them. This can also be in a future PR though.

Thanks again for the great work!

abmfy committed:
Merge branch 'main' into eplb

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Style] Rename `--num-extra-experts` to `--num-redundant-experts`

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Doc] Add glossary for different types of experts

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Doc] Add staatements in `EplbState` that some var is just config

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Doc] Add notes on synchronization of rearrangement step

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Doc] Add examples for expert mappings

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Doc] Add explanation on why picking the last layer for MoE config

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Refactor] Revert `fused_moe.py` since not used

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Doc] Add explanations for calling points of `_dummy_run`

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Doc] Add comments on when do real communication happen

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Doc] Add comments on only last `eplb_window_size` steps will be used

When the rearrangement interval is larger than window size.

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Feature] Disable balancedness logging by default

Since this adds some considerable amount of communication overhead

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Style] Rename shadowed variables to make linter happy

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Style] Add parameters of `apply` for subclasses of `FusedMoEMethodBase`

To make linter happy

Signed-off-by: Bowen Wang <abmfy@icloud.com>

WoosukKwon reviewed (approved):
LGTM! Thanks for the amazing work and let's follow up with more optimizations!

WoosukKwon commented:
@abmfy Oh actually, before we merge this PR, can we have a (unit) test?

abmfy committed:
[Test] Add test for EPLB algo

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy commented:
> Also, maybe not in this PR, but it'd be nice if we can group the eplb-related configs (or ep-related ones) into a separate config. We did it for `compilation_config` and `speculative_config`, and it's quite nice because then we can do `--ep_config={"lb_window_size": 1000, "lb_step_size": 3000, ...}` instead of having a bunch of `--eplb-window-size 1000 --eplb-step-size 3000 --...`.

That's really nice, let me add it to the TODO list

abmfy committed:
[Test] Add test for EPLB execute

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Style] Split some long lines

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
Merge branch 'main' into eplb

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Feature] Use `get_node_count` and remove magic number

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Test] Disable `first_k_dense_replace` in `test_initialization`

Otherwise, there will not be any experts for DeepSeek models,
thus we're unable to get MoE expert information from example MoE layer

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Test] Use only 2 experts in `test_initialization`

For DeepSeek-V3, on the CI env, if we use full experts
this will cause an OOM

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Test] Get at least `n_group` experts in `test_initialization`

Otherwise, topk operation during compilation will fail

Signed-off-by: Bowen Wang <abmfy@icloud.com>

abmfy committed:
[Test] Allow 2 experts per group in `test_initialization`

Since `grouped_topk` will assume top-2 for DeepSeek-V3

Signed-off-by: Bowen Wang <abmfy@icloud.com>

ztxdcyy commented:
🎉 So happy to see this PR finally merged after going through so many challenges — big round of applause for the researcher's persistence and dedication! @abmfy 👏👏👏

Also, just wondering — how can we measure the benefits brought by EPLB? 🤔
Things like expert balancing, GPU utilization, TTFT, TPOT... Any suggestions or best practices? 💡

abmfy commented:
> 🎉 So happy to see this PR finally merged after going through so many challenges — big round of applause for the researcher's persistence and dedication! @abmfy 👏👏👏
> 
> Also, just wondering — how can we measure the benefits brought by EPLB? 🤔 Things like expert balancing, GPU utilization, TTFT, TPOT... Any suggestions or best practices? 💡

Hi @ztxdcyy, thanks for your attention!

There’s now a default-off option `--eplb-log-balancedness` that logs the load balance factor across different GPUs at each step.

As for other metrics, I believe they’re not specific to the EPLB settings, so we can simply rely on standard metrics by running benchmarks and monitoring those results as usual.

Let me know what you think!

abmfy commented:
> @abmfy Hello, I'm encountering the following error when using multi-GPU parallel processing Here's my startup command: python -m vllm.entrypoints.openai.api_server --model="/public/models/hf_models/DeepSeek-V2-Lite-Chat-FP8-A16" --trust-remote-code -tp 2 -dp 2 --port 8200 --enforce-eager --enable-eplb --eplb-log-balancedness
> 
> This issue doesn't occur when starting with a single GPU and only appears during multi-GPU parallel processing. Have you encountered this before, or do you have any solutions?
> 
> ![微信图片_20250704160723](https://private-user-images.githubusercontent.com/82803191/462371528-576ef3ca-406b-45a5-8812-4907d0af3381.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTQwMDA2NTQsIm5iZiI6MTc1NDAwMDM1NCwicGF0aCI6Ii84MjgwMzE5MS80NjIzNzE1MjgtNTc2ZWYzY2EtNDA2Yi00NWE1LTg4MTItNDkwN2QwYWYzMzgxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MzElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzMxVDIyMTkxNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQxZmUxMjlhYzVmYjc0OWNiNmU0M2NhZmM5NTY0MmVkMTE3MDZhYzZjNTRkZmMzODc1NTUyMTI4MGQ3MDczZDYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.aXn9AWQ1LE0wJ3_EBg68AbQi5J5jutgKhuwrMxV4xCI)

Hi @Lichunyan3, sorry for the late reply—I was traveling.

It looks like you may have missed adding `--enable-expert-parallel`; EPLB requires running under EP.

We’ve added some checks in #21102, so if EPLB is enabled without EP, it will now raise an error.

Bounty-hunter commented:
did you test how balancedness imporve in  benchmark_serving.py?  It’s a random dataset. Will there be a significant improvement?
